File: operators/grpc_operators/common/tensor_proto.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights
 * reserved. SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "tensor_proto.hpp"

namespace holoscan::ops {

#ifndef CUDA_TRY
#define CUDA_TRY(stmt)                                                                     \
  ({                                                                                       \
    cudaError_t _holoscan_cuda_err = stmt;                                                 \
    if (cudaSuccess != _holoscan_cuda_err) {                                               \
      GXF_LOG_ERROR("CUDA Runtime call %s in line %d of file %s failed with '%s' (%d).\n", \
                    #stmt,                                                                 \
                    __LINE__,                                                              \
                    __FILE__,                                                              \
                    cudaGetErrorString(_holoscan_cuda_err),                                \
                    _holoscan_cuda_err);                                                   \
    }                                                                                      \
    _holoscan_cuda_err;                                                                    \
  })
#endif

void TensorProto::gxf_time_to_proto(const nvidia::gxf::Entity& gxf_entity,
                                    ::holoscan::entity::Timestamp* timestamp) {
  auto gxf_timestamp = gxf_entity.get<nvidia::gxf::Timestamp>();
  if (gxf_timestamp) {
    timestamp->set_acqtime((*gxf_timestamp)->acqtime);
    timestamp->set_pubtime((*gxf_timestamp)->pubtime);
  }
}

void TensorProto::proto_to_gxf_time(nvidia::gxf::Entity& gxf_entity,
                                    const ::holoscan::entity::Timestamp& timestamp) {
  auto gxf_timestamp = gxf_entity.add<nvidia::gxf::Timestamp>("timestamp");
  (*gxf_timestamp)->acqtime = timestamp.acqtime();
  (*gxf_timestamp)->pubtime = timestamp.pubtime();
}

void TensorProto::gxf_tensor_to_proto(
    const nvidia::gxf::Entity& gxf_entity,
    google::protobuf::Map<std::string, ::holoscan::entity::Tensor>* tensor_map,
    const cudaStream_t cuda_stream) {
  auto tensors = gxf_entity.findAll<nvidia::gxf::Tensor, 4>();
  if (!tensors) { throw std::runtime_error("Tensor not found"); }

  for (auto tensor : tensors.value()) {
    holoscan::entity::Tensor& tensor_proto = (*tensor_map)[tensor->name()];
    for (uint32_t i = 0; i < (*tensor)->shape().rank(); i++) {
      tensor_proto.add_dimensions((*tensor)->shape().dimension(i));
    }
    switch ((*tensor)->element_type()) {
      case nvidia::gxf::PrimitiveType::kUnsigned8:
        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kUnsigned8);
        copy_data_to_proto<uint8_t>(tensor.value(), tensor_proto, cuda_stream);
        break;
      case nvidia::gxf::PrimitiveType::kUnsigned16:
        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kUnsigned16);
        copy_data_to_proto<uint16_t>(tensor.value(), tensor_proto, cuda_stream);
        break;
      case nvidia::gxf::PrimitiveType::kFloat32:
        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kFloat32);
        copy_data_to_proto<float>(tensor.value(), tensor_proto, cuda_stream);
        break;
      default:
        throw std::runtime_error("Unsupported primitive type");
    }
  }
}

template <typename T>
void TensorProto::copy_data_to_proto(const nvidia::gxf::Handle<nvidia::gxf::Tensor>& tensor,
                                     ::holoscan::entity::Tensor& tensor_proto,
                                     const cudaStream_t cuda_stream) {
  if ((*tensor).storage_type() == nvidia::gxf::MemoryStorageType::kDevice) {
    void* in_data_ptr = (*tensor).pointer();
    size_t data_size = (*tensor).bytes_size();
    std::vector<T> in_data(data_size);
    CUDA_TRY(cudaMemcpyAsync(
        in_data.data(), in_data_ptr, data_size, cudaMemcpyDeviceToHost, cuda_stream));
    tensor_proto.set_data(in_data.data(), data_size);
    tensor_proto.set_memory_storage_type(holoscan::entity::Tensor::kDevice);
  } else {
    tensor_proto.set_data((*tensor).pointer(), (*tensor).size());
    tensor_proto.set_memory_storage_type(holoscan::entity::Tensor::kHost);
  }
}

void TensorProto::proto_to_gxf_tensor(
    nvidia::gxf::Entity& gxf_entity,
    const google::protobuf::Map<std::string, ::holoscan::entity::Tensor>& tensor_map,
    nvidia::gxf::Handle<nvidia::gxf::Allocator>& allocator, const cudaStream_t cuda_stream) {
  for (auto tensor_entry : tensor_map) {
    const holoscan::entity::Tensor& tensor_proto = tensor_entry.second;
    auto tensor = gxf_entity.add<nvidia::gxf::Tensor>(tensor_entry.first.c_str());
    if (!tensor) { throw std::runtime_error("Failed to create tensor"); }

    nvidia::gxf::Shape shape({tensor_proto.dimensions().begin(), tensor_proto.dimensions().end()});
    switch (tensor_proto.primitive_type()) {
      case holoscan::entity::Tensor::kUnsigned8:
        tensor.value()->reshape<uint8_t>(
            shape, memory_storage_type(tensor_proto.memory_storage_type()), allocator);
        copy_data_to_tensor<uint8_t>(tensor_proto, tensor.value(), cuda_stream);
        break;
      case holoscan::entity::Tensor::kUnsigned16:
        tensor.value()->reshape<uint16_t>(
            shape, memory_storage_type(tensor_proto.memory_storage_type()), allocator);
        copy_data_to_tensor<uint16_t>(tensor_proto, tensor.value(), cuda_stream);
        break;
      case holoscan::entity::Tensor::kFloat32:
        tensor.value()->reshape<float>(
            shape, memory_storage_type(tensor_proto.memory_storage_type()), allocator);
        copy_data_to_tensor<float>(tensor_proto, tensor.value(), cuda_stream);
        break;
      default:
        throw std::runtime_error("Unsupported primitive type");
    }
  }
}
nvidia::gxf::MemoryStorageType TensorProto::memory_storage_type(
    holoscan::entity::Tensor_MemoryStorageType mem_storage_type) {
  switch (mem_storage_type) {
    case holoscan::entity::Tensor::kDevice:
      return nvidia::gxf::MemoryStorageType::kDevice;
    case holoscan::entity::Tensor::kHost:
      return nvidia::gxf::MemoryStorageType::kHost;
    case holoscan::entity::Tensor::kSystem:
      return nvidia::gxf::MemoryStorageType::kSystem;
    default:
      throw std::runtime_error("Unsupported memory storage type");
  }
}

template <typename T>
void TensorProto::copy_data_to_tensor(const ::holoscan::entity::Tensor& tensor_proto,
                                      nvidia::gxf::Handle<nvidia::gxf::Tensor>& tensor,
                                      const cudaStream_t cuda_stream) {
  if (tensor_proto.memory_storage_type() == holoscan::entity::Tensor::kDevice) {
    CUDA_TRY(cudaMemcpyAsync((*tensor).pointer(),
                             tensor_proto.data().data(),
                             tensor_proto.data().size(),
                             cudaMemcpyHostToDevice,
                             cuda_stream));
  } else {
    std::copy(tensor_proto.data().begin(), tensor_proto.data().end(), (*tensor).pointer());
  }
}

void TensorProto::tensor_to_entity_request(const nvidia::gxf::Entity& gxf_entity,
                                           std::shared_ptr<EntityRequest> request,
                                           const cudaStream_t cuda_stream) {
  TensorProto::gxf_time_to_proto(gxf_entity, request->mutable_timestamp());
  TensorProto::gxf_tensor_to_proto(gxf_entity, request->mutable_tensors(), cuda_stream);
}

void TensorProto::tensor_to_entity_response(const nvidia::gxf::Entity& gxf_entity,
                                            std::shared_ptr<EntityResponse> response,
                                            const cudaStream_t cuda_stream) {
  TensorProto::gxf_time_to_proto(gxf_entity, response->mutable_timestamp());
  TensorProto::gxf_tensor_to_proto(gxf_entity, response->mutable_tensors(), cuda_stream);
}

void TensorProto::entity_request_to_tensor(
    const EntityRequest* entity_request, nvidia::gxf::Entity& gxf_entity,
    nvidia::gxf::Handle<nvidia::gxf::Allocator> gxf_allocator, const cudaStream_t cuda_stream) {
  TensorProto::proto_to_gxf_time(gxf_entity, entity_request->timestamp());
  TensorProto::proto_to_gxf_tensor(
      gxf_entity, entity_request->tensors(), gxf_allocator, cuda_stream);
}

void TensorProto::entity_response_to_tensor(
    const EntityResponse& entity_response, nvidia::gxf::Entity& gxf_entity,
    nvidia::gxf::Handle<nvidia::gxf::Allocator> gxf_allocator, const cudaStream_t cuda_stream) {
  TensorProto::proto_to_gxf_time(gxf_entity, entity_response.timestamp());
  TensorProto::proto_to_gxf_tensor(
      gxf_entity, entity_response.tensors(), gxf_allocator, cuda_stream);
}
}  // namespace holoscan::ops

File: operators/grpc_operators/python/common/tensor_proto.py
# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import logging
import sys
from enum import Enum
from typing import Dict, Union

import cupy as cp
import numpy as np
from holoscan import as_tensor
from holoscan.core import DLDeviceType
from holoscan.gxf import Entity

from holohub.grpc_operators import holoscan_pb2 as holoscan_proto

logger = logging.getLogger(__name__)

current_module = sys.modules[__name__]


class ProtoTensorCudaArrayInterface:
    def __init__(self, proto_tensor):
        data = np.frombuffer(proto_tensor.data, dtype=TensorProto._get_numpy_type(proto_tensor))
        data = cp.asarray(data)
        self.__cuda_array_interface__ = {
            "version": 2,
            "shape": tuple(proto_tensor.dimensions),
            "data": (data.data.ptr, False),
            "typestr": proto_tensor.type_str,
            "strides": tuple(proto_tensor.strides),
        }


class DLDataTypeCode(Enum):
    kDLInt = 0
    kDLUInt = 1
    kDLFloat = 2
    kDLOpaqueHandle = 3
    kDLComplex = 5


class TensorProto:
    NULL_KEY_NAME = "__NULL__"

    STORAGE_TYPES = {
        holoscan_proto.Tensor.MemoryStorageType.kHost: DLDeviceType.DLCPU,
        holoscan_proto.Tensor.MemoryStorageType.kDevice: DLDeviceType.DLCUDA,
        holoscan_proto.Tensor.MemoryStorageType.kSystem: DLDeviceType.DLCUDAHOST,
    }

    PRIMITIVE_TYPES = {
        holoscan_proto.Tensor.PrimitiveType.kUnsigned8: np.uint8,
        holoscan_proto.Tensor.PrimitiveType.kUnsigned16: np.uint16,
        holoscan_proto.Tensor.PrimitiveType.kFloat32: np.float32,
    }

    @staticmethod
    def add_tensor_to_proto(gxf_entity, response: holoscan_proto.EntityResponse):
        TensorProto._gxf_tensors_to_proto_tensors(gxf_entity, response.tensors)

    @staticmethod
    def tensor_to_proto(gxf_entity: Dict) -> holoscan_proto.EntityRequest:
        entity_request = holoscan_proto.EntityRequest()
        TensorProto._gxf_tensors_to_proto_tensors(gxf_entity, entity_request.tensors)
        return entity_request

    @staticmethod
    def proto_to_tensor(
        entity_request: Union[holoscan_proto.EntityRequest, holoscan_proto.EntityResponse], context
    ) -> Entity:
        gxf_entity = Entity(context)
        TensorProto._proto_tensors_to_gxf_tensors(entity_request, gxf_entity)
        return gxf_entity

    @staticmethod
    def _gxf_tensors_to_proto_tensors(gxf_entity, tensors):
        for key, gxf_tensor in gxf_entity.items():
            try:
                tensor = tensors[key if key else TensorProto.NULL_KEY_NAME]

                for dim in gxf_tensor.shape:
                    tensor.dimensions.append(dim)

                for stride in gxf_tensor.strides:
                    tensor.strides.append(stride)

                tensor.primitive_type = TensorProto._get_gxf_type(gxf_tensor)
                tensor.type_str = TensorProto._get_numpy_type_str(gxf_tensor)
                tensor.memory_storage_type = TensorProto._get_proto_storage_type(gxf_tensor)

                if (
                    tensor.memory_storage_type == holoscan_proto.Tensor.MemoryStorageType.kHost
                    or tensor.memory_storage_type == holoscan_proto.Tensor.MemoryStorageType.kSystem
                ):
                    np_array = np.asarray(
                        gxf_tensor, dtype=TensorProto.PRIMITIVE_TYPES[tensor.primitive_type]
                    )
                    tensor.data = np_array.tobytes()
                elif tensor.memory_storage_type == holoscan_proto.Tensor.MemoryStorageType.kDevice:
                    cp_array = cp.asarray(
                        gxf_tensor, dtype=TensorProto.PRIMITIVE_TYPES[tensor.primitive_type]
                    )
                    tensor.data = cp.asnumpy(cp_array).tobytes()

            except Exception as e:
                logger.error(f"Failed to convert tensor with key {key}: {str(e)}")
                raise

    @staticmethod
    def _proto_tensors_to_gxf_tensors(entity_request, gxf_entity):
        for key, proto_tensor in entity_request.tensors.items():
            try:
                dtype = TensorProto._get_numpy_type(proto_tensor)

                if (
                    proto_tensor.memory_storage_type
                    == holoscan_proto.Tensor.MemoryStorageType.kHost
                    or proto_tensor.memory_storage_type
                    == holoscan_proto.Tensor.MemoryStorageType.kSystem
                ):
                    tensor = as_tensor(
                        np.frombuffer(proto_tensor.data, dtype=dtype)
                        .reshape(proto_tensor.dimensions)
                        .copy()
                    )
                elif (
                    proto_tensor.memory_storage_type
                    == holoscan_proto.Tensor.MemoryStorageType.kDevice
                ):
                    tensor = as_tensor(cp.asarray(ProtoTensorCudaArrayInterface(proto_tensor)))

                gxf_entity.add(tensor, key if key != TensorProto.NULL_KEY_NAME else "")
            except Exception as e:
                logger.error(f"Failed to convert tensor with key {key}: {str(e)}")
                raise

    @staticmethod
    def _get_gxf_storage_type(proto_tensor):
        if proto_tensor.memory_storage_type in TensorProto.STORAGE_TYPES:
            return TensorProto.STORAGE_TYPES[proto_tensor.memory_storage_type]
        raise ValueError(f"Unsupported memory storage type: {proto_tensor.memory_storage_type}")

    @staticmethod
    def _get_numpy_type(proto_tensor):

        if proto_tensor.primitive_type in TensorProto.PRIMITIVE_TYPES:
            return TensorProto.PRIMITIVE_TYPES[proto_tensor.primitive_type]

        raise ValueError(f"Unsupported primitive type: {proto_tensor.primitive_type}")

    @staticmethod
    def _get_proto_storage_type(gxf_tensor):
        if gxf_tensor.device.device_type in TensorProto.STORAGE_TYPES.values():
            return next(
                key
                for key, value in TensorProto.STORAGE_TYPES.items()
                if value == gxf_tensor.device.device_type
            )

        raise ValueError(f"Unsupported device type: {gxf_tensor.device.device_type}")

    @staticmethod
    def _get_gxf_type(value):
        if value.dtype.code == DLDataTypeCode.kDLUInt.value:
            if value.dtype.bits == 8:
                return holoscan_proto.Tensor.PrimitiveType.kUnsigned8
            if value.dtype.bits == 16:
                return holoscan_proto.Tensor.PrimitiveType.kUnsigned16
            if value.dtype.bits == 32:
                return holoscan_proto.Tensor.PrimitiveType.kUnsigned32
            if value.dtype.bits == 64:
                return holoscan_proto.Tensor.PrimitiveType.kUnsigned64
        elif value.dtype.code == DLDataTypeCode.kDLFloat.value:
            if value.dtype.bits == 16:
                return holoscan_proto.Tensor.PrimitiveType.kFloat16
            elif value.dtype.bits == 32:
                return holoscan_proto.Tensor.PrimitiveType.kFloat32
            elif value.dtype.bits == 64:
                return holoscan_proto.Tensor.PrimitiveType.kFloat64

        raise ValueError(f"Unsupported primitive type: {value.dtype}")

    @staticmethod
    def _get_numpy_type_str(gxf_tensor):
        if gxf_tensor.dtype.lanes != 1:
            raise ValueError(
                "typestring conversion only support DLDataType with one lane, "
                f"but found dtype.lanes: ({gxf_tensor.dtype.lanes})."
            )

        if gxf_tensor.dtype.code == DLDataTypeCode.kDLUInt.value:
            if gxf_tensor.dtype.bits == 8:
                return "|u1"
            if gxf_tensor.dtype.bits == 16:
                return "<u2"
            if gxf_tensor.dtype.bits == 32:
                return "<u4"
            if gxf_tensor.dtype.bits == 64:
                return "<u8"
        elif gxf_tensor.dtype.code == DLDataTypeCode.kDLFloat.value:
            if gxf_tensor.dtype.bits == 16:
                return "<f2"
            elif gxf_tensor.dtype.bits == 32:
                return "<f4"
            elif gxf_tensor.dtype.bits == 64:
                return "<f8"

        raise ValueError(f"Data type code ({gxf_tensor.dtype.code}) is not supported!")
